Android/app/src/go/intra/split/retrier.go

```diff
@@ -219,22 +219,75 @@ func (r *retrier) CloseRead() error {
 	return r.conn.CloseRead()
 }
 
-func splitHello(hello []byte) ([]byte, []byte) {
-	if len(hello) == 0 {
-		return hello, hello
+func getTLSClientHelloRecordLen(h []byte) (uint16, bool) {
+	if len(h) < 5 {
+		return 0, false
+	}
+
+	const (
+		TYPE_HANDSHAKE byte   = 22
+		VERSION_TLS10  uint16 = 0x0301
+		VERSION_TLS11  uint16 = 0x0302
+		VERSION_TLS12  uint16 = 0x0303
+		VERSION_TLS13  uint16 = 0x0304
+	)
+
+	if h[0] != TYPE_HANDSHAKE {
+		return 0, false
+	}
+
+	ver := binary.BigEndian.Uint16(h[1:3])
+	if ver != VERSION_TLS10 && ver != VERSION_TLS11 &&
+		ver != VERSION_TLS12 && ver != VERSION_TLS13 {
+		return 0, false
+	}
+
+	return binary.BigEndian.Uint16(h[3:5]), true
+}
+
+func splitHello(hello []byte, w io.Writer) (n int, splitLen int, err error) {
+	if len(hello) <= 1 {
+		n, err = w.Write(hello)
+		return
 	}
+
 	const (
-		MIN_SPLIT int = 32
+		MIN_SPLIT int = 6
 		MAX_SPLIT int = 64
 	)
 
 	// Random number in the range [MIN_SPLIT, MAX_SPLIT]
-	s := MIN_SPLIT + rand.Intn(MAX_SPLIT+1-MIN_SPLIT)
+	// splitLen includes 5 bytes of TLS header
+	splitLen = MIN_SPLIT + rand.Intn(MAX_SPLIT+1-MIN_SPLIT)
 	limit := len(hello) / 2
-	if s > limit {
-		s = limit
+	if splitLen > limit {
+		splitLen = limit
 	}
-	return hello[:s], hello[s:]
+
+	recordLen, ok := getTLSClientHelloRecordLen(hello)
+	recordSplitLen := splitLen - 5
+	if !ok || recordSplitLen <= 0 || recordSplitLen >= int(recordLen) {
+		// Do TCP split if hello is not a valid TLS Client Hello, or cannot be fragmented
+		n, err = w.Write(hello[:splitLen])
+		if err == nil {
+			var m int
+			m, err = w.Write(hello[splitLen:])
+			n += m
+		}
+		return
+	}
+
+	binary.BigEndian.PutUint16(hello[3:5], uint16(recordSplitLen))
+	if n, err = w.Write(hello[:splitLen]); err != nil {
+		return
+	}
+
+	copy(hello[splitLen-5:splitLen], hello[:5])
+	binary.BigEndian.PutUint16(hello[splitLen-2:splitLen], recordLen-uint16(recordSplitLen))
+	var m int
+	m, err = w.Write(hello[splitLen-5:])
```

Comment:
Assign each slice to a `pkt` slice and the code becomes clearer because the indexes are simplified:

```suggestion
	pkt := hello[:splitLen]
	binary.BigEndian.PutUint16(pkt[3:5], uint16(recordSplitLen))
	if n, err = w.Write(pkt); err != nil {
		return
	}

	pkt = hello[splitLen-5:]
	copy(pkt, hello[:5])
 	binary.BigEndian.PutUint16(pkt[3:5], recordLen-uint16(recordSplitLen))
	var m int
	m, err = w.Write(pkt)
```